### PR TITLE
feat(ingest): redshift - Redshift rework

### PR DESCRIPTION
**Pull request by treff7es**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

**Comment by treff7es** [comment]
CleanShot 2023-01-26 at 18 05 26@2x
CleanShot 2023-01-26 at 18 05 18@2x
CleanShot 2023-01-26 at 18 05 08@2x
CleanShot 2023-01-26 at 18 04 49@2x
CleanShot 2023-01-26 at 18 04 38@2x
CleanShot 2023-01-26 at 18 04 30@2x
CleanShot 2023-01-26 at 18 04 17@2x
CleanShot 2023-01-26 at 18 04 05@2x

**Review by hsheth2 — DISMISSED** [review]
Took a cursory look over this so far

**Comment by treff7es** [comment]
I added complex type support as well:
CleanShot 2023-02-06 at 10 35 47@2x

File: metadata-ingestion/setup.py
Comment (anshbansal) [inline metadata-ingestion/setup.py:536]: breaking change. Please see if we can avoid it. Also, need to update [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md) file

File: metadata-ingestion/src/datahub/ingestion/source/redshift/lineage.py
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/lineage.py:352]: can we locate these inside a queries.py constants file?

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:254]: Can just return `not` this expression, or the demorgan's law equivalent

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:301]: Is this completed?

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:306]: Is this purposefully not yielding to fetch from the db all at once? The comment and method name make it sound like it should be a generator

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:330]: Think is checked twice, once here and once in `_gen_access_events_from_history_query`

**Review by asikowitz — COMMENTED** [review]
Sorry for the incredibly long review. Comments range from typos and very minor style suggestions, to larger refactoring suggestions, to plain questions. Kinda glad I'm picking this up because this is an overwhelming number of comments

File: metadata-ingestion/src/datahub/ingestion/source/redshift/redshift.py
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/redshift.py:899]: Can use a `yield from`; same below

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py
Comment (asikowitz) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:288]: Is this very common? Seems like something we should report in non-debug mode

**Review by asikowitz — COMMENTED** [review]
Really nice job on refactoring some of the more confusing code. I have a few more comments, but we should bias towards getting this in when it's in a good state over doing more style changes. Holding off approval until a merge is completed because there are some potentially breaking changes, e.g. around merging the sub type changes

**Review by asikowitz — APPROVED** [review]
LGTM! Thanks for working through the many requested changes

File: metadata-ingestion/src/datahub/ingestion/source/redshift/redshift_schema.py
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/redshift/redshift_schema.py:175]: These fields are listed in the SQL query, so they should be available. The only issue can be if you delete/rename in the query but I think we should fail in that case

File: metadata-ingestion/src/datahub/ingestion/source/redshift/lineage.py
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/lineage.py:213]: much nicer!

File: metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/redshift/usage.py:347]: nice!

File: metadata-ingestion/src/datahub/specific/dataset.py
Comment (treff7es) [inline metadata-ingestion/src/datahub/specific/dataset.py:117]: @hsheth2 This breaks integration tests as well but I had to do it as `s3` path with `/` caused lineage to break.

Comment (hsheth2) [reply metadata-ingestion/src/datahub/specific/dataset.py:117]: Can we change this to exclusively quote `/` characters but not need to quote everything?

Also did you test how this works with our actual patch applier?

**Comment by hsheth2** [comment]
@treff7es looks like the docs site still doesn't like the redshift-usage-legacy fixes

```
Failed to process redshift-usage-legacy due to exception No module named 'datahub.ingestion.source.usage.redshift_usage'
```

**Pull request merged**
